File: ios/DayAtlas/Location/TrackingController.swift
import Foundation

/// Mirrors app/src/main/java/com/dayatlas/app/location/TrackingController.kt.
enum TrackingController {
    /// Called from AppDelegate.didFinishLaunchingWithOptions on every launch,
    /// including a silent relaunch that iOS performs to deliver a
    /// significant-location-change event (which is what can happen right
    /// after the device reboots, if the user already granted "Always" and
    /// hasn't force-quit the app since).
    static func onAppLaunch() {
        let prefs = AppPrefs.shared
        if prefs.dailyMode {
            prefs.trackingEnabled = true
        }
        if prefs.trackingEnabled {
            LocationTracker.shared.resume()
        }
    }

    static func start(sampleSoon: Bool = true) {
        AppPrefs.shared.trackingEnabled = true
        LocationTracker.shared.resume()
        if sampleSoon {
            LocationTracker.shared.sampleNow()
        }
    }

    static func stop() {
        AppPrefs.shared.trackingEnabled = false
        LocationTracker.shared.pause()
    }

    static func setDailyMode(_ enabled: Bool) {
        AppPrefs.shared.dailyMode = enabled
        if enabled {
            start(sampleSoon: true)
        } else {
            stop()
        }
    }
}

File: app/src/main/java/com/dayatlas/app/location/TrackingController.kt
package com.dayatlas.app.location

import android.content.Context
import com.dayatlas.app.prefs.AppPrefs

object TrackingController {
    fun onAppStart(context: Context, prefs: AppPrefs = AppPrefs(context)) {
        if (prefs.dailyMode) {
            prefs.trackingEnabled = true
        }
        if (prefs.trackingEnabled) {
            SampleScheduler.ensureScheduled(context, prefs)
        }
    }

    fun start(context: Context, prefs: AppPrefs = AppPrefs(context), sampleSoon: Boolean = true) {
        prefs.trackingEnabled = true
        prefs.resetStationaryBackoff()
        if (sampleSoon) {
            SampleScheduler.scheduleNext(context, prefs, delayMs = 3_000L)
        } else {
            SampleScheduler.ensureScheduled(context, prefs)
        }
    }

    fun stop(context: Context, prefs: AppPrefs = AppPrefs(context)) {
        prefs.trackingEnabled = false
        prefs.resetStationaryBackoff()
        SampleScheduler.cancel(context)
    }

    fun setDailyMode(context: Context, enabled: Boolean, prefs: AppPrefs = AppPrefs(context)) {
        prefs.dailyMode = enabled
        if (enabled) {
            start(context, prefs, sampleSoon = true)
        } else {
            stop(context, prefs)
        }
    }

    fun shouldSample(context: Context, prefs: AppPrefs = AppPrefs(context)): Boolean {
        if (!prefs.trackingEnabled && !prefs.dailyMode) return false
        return PermissionHelper.hasLocation(context)
    }
}
